Sources/Foundation/Decimal.swift

```diff
@@ -718,22 +812,47 @@ extension Decimal {
     }
 
     public var ulp: Decimal {
-        if !self.isFinite { return Decimal.nan }
+        guard isFinite else { return .nan }
+
+        let exponent: Int32
+        if isZero {
+            exponent = .min
+        } else {
+            let significand_ = significand
+            let shift =
+                _powersOfTenDividingUInt128Max.firstIndex { significand_ > $0 }
+                    ?? _powersOfTenDividingUInt128Max.count
+            exponent = _exponent &- Int32(shift)
+        }
+
         return Decimal(
-            _exponent: _exponent, _length: 1, _isNegative: 0, _isCompact: 1,
+            _exponent: max(exponent, -128), _length: 1, _isNegative: 0, _isCompact: 1,
             _reserved: 0, _mantissa: (0x0001, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000))
     }
 
     public var nextUp: Decimal {
-        return self + Decimal(
-            _exponent: _exponent, _length: 1, _isNegative: 0, _isCompact: 1,
-            _reserved: 0, _mantissa: (0x0001, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000))
+        if _isNegative == 1 {
```

Comment:
This logic is the `Decimal` counterpart to https://github.com/apple/swift/blob/4c923d6dda6453d3af635f0ac1d64935a21673bb/stdlib/public/core/FloatingPointTypes.swift.gyb#L946.